### PR TITLE
[Feature] dist op role and lr op role, to support memory optimize with dist training

### PR DESCRIPTION
**Pull request by typhoonzero**
Should Fix https://github.com/PaddlePaddle/Paddle/issues/13042

在修复 https://github.com/PaddlePaddle/Paddle/issues/13042 的过程中和 @Yancey1989 发现的一个可以简化代码的方法

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_dist_word2vec.py [hunk 1]
```diff
@@ -17,15 +17,24 @@
...
         self.check_with_place("dist_word2vec.py", delta=1e-4)
 
 
-class TestDistSeResneXt2x2Async(TestDistBase):
+class TestDistW2V2x2WithMemOpt(TestDistBase):
+    def _setup_config(self):
+        self._sync_mode = True
+        self._mem_opt = True
+
+    def test_se_resnext(self):
```
Comment (Yancey0623) [inline python/paddle/fluid/tests/unittests/test_dist_word2vec.py]: test_word2vec(self) ?

**Comment by typhoonzero** [comment]
This fix will depend on memory_optimization transpile should output same program when input program are the same, this will cause distribute transpiler output different program on different nodes.

File: python/paddle/fluid/transpiler/details/program_utils.py [hunk 2]
```diff
@@ -21,13 +21,17 @@
...
+        try:
+            idx = list(block.ops).index(op)
+            block._remove_op(idx)
+            # del block.ops[idx]
+            # end = list(block.ops).index(ops[-1])
+            # [block._remove_op(start) for _ in six.moves.range(end - start + 1)]
+        except Exception as e:
+            print(e)
+        # raise e
+        # block.program._sync_with_cpp()
```
Comment (Yancey0623) [inline python/paddle/fluid/transpiler/details/program_utils.py]: Do we need call the `sync_with_cpp()` function? And please delete the unused code.

**Review by Yancey0623 — DISMISSED** [review]
LGTM, to speed up merging, we can fix the comment in another PR.

**Review by Yancey0623 — APPROVED** [review]
LGTM.

**Pull request merged**